experimental/ir/lower_options.go

```diff
@@ -0,0 +1,351 @@
+package ir
+
+import (
+	"fmt"
+	"iter"
+	"slices"
+
+	"github.com/bufbuild/protocompile/experimental/ast"
+	"github.com/bufbuild/protocompile/experimental/internal/taxa"
+	"github.com/bufbuild/protocompile/experimental/ir/presence"
+	"github.com/bufbuild/protocompile/experimental/report"
+	"github.com/bufbuild/protocompile/experimental/seq"
+	"github.com/bufbuild/protocompile/internal/arena"
+	"github.com/bufbuild/protocompile/internal/ext/iterx"
+)
+
+// resolveOptions resolves all of the options in a file.
+func resolveOptions(f File, r *report.Report) {
+	dpIdx := int32(len(f.Context().imports.files))
+	dp := f.Context().imports.DescriptorProto().Context()
+	fileOptions := ref[rawField]{dpIdx, dp.langSymbols.fileOptions}
+	messageOptions := ref[rawField]{dpIdx, dp.langSymbols.messageOptions}
+	fieldOptions := ref[rawField]{dpIdx, dp.langSymbols.fieldOptions}
+	oneofOptions := ref[rawField]{dpIdx, dp.langSymbols.oneofOptions}
+	enumOptions := ref[rawField]{dpIdx, dp.langSymbols.enumOptions}
+	enumValueOptions := ref[rawField]{dpIdx, dp.langSymbols.enumValueOptions}
+
+	bodyOptions := func(b ast.DeclBody) iter.Seq[ast.Option] {
+		return iterx.FilterMap(seq.Values(b.Decls()), func(d ast.DeclAny) (ast.Option, bool) {
+			def := d.AsDef()
+			if def.IsZero() || def.Classify() != ast.DefKindOption {
+				return ast.Option{}, false
+			}
+			return def.AsOption().Option, true
+		})
+	}
+
+	for def := range bodyOptions(f.AST().DeclBody) {
+		optionRef{
+			Context: f.Context(),
+			Report:  r,
+
+			scope: f.Package(),
+			def:   def,
+
+			field: fileOptions,
+			raw:   &f.Context().options,
+		}.resolve()
+	}
+
+	for ty := range seq.Values(f.AllTypes()) {
+		for ast := range bodyOptions(ty.AST().Body()) {
```

Comment:
Maybe name this `def` instead of `ast`, for consistency with the loop above and the field name in `optionRef` (and also so it doesn't collide with package name).

Here and below/throughout.